Add the `Symfony\Config\config()` function

alexandre-daubois opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | -
| License       | MIT

This PR adds a new `config()` function to configure a Symfony application:

```php
<?php

use function Symfony\Config\config;

return config(
    framework: [
        'cache' => [
            'prefix_seed' => 'seed_'.$this->env.'_',
        ],
    ],
    security: [
        'access_decision_manager' => [
            'strategy' => 'unanimous',
        ],
    ]
);
```

The fonction is generated at cache warmup, allowing to generate the corresponding phpdoc along with Jetbrain's `ArrayShape` attribute. Here a sneak pick of what the function looks like:

```php
<?php

namespace Symfony\Config;

use Jetbrains\PhpStorm\ArrayShape;

/**
 * This class is automatically generated to help in creating a config.
 *
 * @param array{
 *     secret?: string|int|float|bool,
 *     http_method_override?: bool,
 *     trust_x_sendfile_type_header?: string|int|float|bool,
 *     ide?: string|int|float|bool,
 *     test?: bool,
 *     ...
 *     form?: array{
 *         enabled?: bool,
 *         csrf_protection?: array{
 *             enabled?: string|int|float|bool,
 *             token_id?: string|int|float|bool,
 *             field_name?: string|int|float|bool,
 *             field_attr?: array<array-key, mixed>,
 *         },
 *     },
 *     ...
 *   } $framework
 *
 * @param array{
 *     access_denied_url?: string|int|float|bool,
 *     session_fixation_strategy?: "none"|"migrate"|"invalidate",
 *     hide_user_not_found?: bool,
 *     erase_credentials?: bool,
 *     access_decision_manager?: array{
 *         strategy?: "affirmative"|"consensus"|"unanimous"|"priority",
 *         service?: string|int|float|bool,
 *         strategy_service?: string|int|float|bool,
 *         allow_if_all_abstain?: bool,
 *         allow_if_equal_granted_denied?: bool,
 *     },
 *     ...
 *   } $security
 */
function config(
    // framework:
    #[ArrayShape([
        'secret' => 'string|int|float|bool',
        'http_method_override' => 'bool', /* Set true to enable support for the '_method' request parameter to determine the intended HTTP method on POST requests. Note: When using the HttpCache, you need to call the method in your front controller instead. Default value: false. */
        'trust_x_sendfile_type_header' => 'string|int|float|bool', /* Set true to enable support for xsendfile in binary file responses. Default value: "%env(bool:default::SYMFONY_TRUST_X_SENDFILE_TYPE_HEADER)%". */
        'ide' => 'string|int|float|bool', /* Default value: "%env(default::SYMFONY_IDE)%". */
        'test' => 'bool',
        // ...
        'form' => [
            'enabled' => 'bool', /* Default value: false. */
            'csrf_protection' => [
                'enabled' => 'string|int|float|bool', /* Default value: null. */
                'token_id' => 'string|int|float|bool', /* Default value: null. */
                'field_name' => 'string|int|float|bool', /* Default value: "_token". */
                'field_attr' => 'array<array-key, mixed>',
            ],
    ], /* Form configuration */
    // ...
    ])] array $framework = [],
    // security:
    #[ArrayShape([
        'access_denied_url' => 'string|int|float|bool', /* Default value: null. */
        'session_fixation_strategy' => '"none"|"migrate"|"invalidate"', /* Default value: "migrate". */
        'hide_user_not_found' => 'bool', /* Default value: true. */
        'erase_credentials' => 'bool', /* Default value: true. */
        'access_decision_manager' => [
            'strategy' => '"affirmative"|"consensus"|"unanimous"|"priority"',
            'service' => 'string|int|float|bool',
            'strategy_service' => 'string|int|float|bool',
            'allow_if_all_abstain' => 'bool', /* Default value: false. */
            'allow_if_equal_granted_denied' => 'bool', /* Default value: true. */
        ],
        // ...
    ])] array $security = []): \Closure {
    // ...
}
```

alamirault committed:
[Routing] Rename annotations to attribute in `AttributeClassLoader`

alamirault committed:
[Serializer] Rename annotations to attributes in AttributeLoader

alexandre-daubois committed:
[FrameworkBundle] Add the `config()` function

nicolas-grekas commented on src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php:
```diff
@@ -0,0 +1,182 @@
+final class ArrayShapeGenerator
+{
+    public const FORMAT_PHPDOC = 'phpdoc';
+    public const FORMAT_JETBRAINS_ARRAY_SHAPE = 'jetbrains_array_shape';
```
```suggestion
    public const FORMAT_JETBRAINS_ATTRIBUTE = 'jetbrains_attribute';
```

nicolas-grekas commented on src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php:
```diff
@@ -0,0 +1,182 @@
+
+    private static function doGeneratePhpDoc(NodeInterface $node, int $nestingLevel = 1): string
+    {
+        if ($node instanceof ArrayNode) {
```
```suggestion
        if (!$node instanceof ArrayNode) { return
```

nicolas-grekas commented on src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php:
```diff
@@ -0,0 +1,182 @@
+    private static function doGeneratePhpDoc(NodeInterface $node, int $nestingLevel = 1): string
+    {
+        if ($node instanceof ArrayNode) {
+            $children = $node->getChildren();
+            $arrayShape = 'array';
+
+            if (!$children) {
```
```suggestion
            $arrayShape = 'array';

            if (!$children = $node->getChildren()) {
```

nicolas-grekas commented on src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php:
```diff
@@ -0,0 +1,182 @@
+
+    private static function doGenerateJetBrainsArrayShape(NodeInterface $node, int $nestingLevel = 1): string
+    {
+        if ($node instanceof ArrayNode) {
```
```suggestion
        if (!$node instanceof ArrayNode) {return
```

nicolas-grekas commented on src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php:
```diff
@@ -0,0 +1,182 @@
+            $shape = '';
+            if (1 === $nestingLevel) {
+                $shape = '#[ArrayShape(';
+            } else {
+                if (!$children) {
```
```suggestion
            } elseif (!$children) {
```

nicolas-grekas commented on src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php:
```diff
@@ -0,0 +1,182 @@
+            $type = 'int<%s, %s>|float<%1$s, %2$s>';
+        }
+
+        $min = null !== $node->getMin() ? $node->getMin() : 'min';
+        $max = null !== $node->getMax() ? $node->getMax() : 'max';
```
```suggestion
        $min = $node->getMin() ?? 'min';
        $max = $node->getMax() ?? 'max';
```

nicolas-grekas commented on src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php:
```diff
@@ -0,0 +1,182 @@
+
+    private static function prependPhpDocWithStar(string $shape): string
+    {
+        return strtr($shape, ["\n" => "\n * "]);
```
str_replace

nicolas-grekas commented on src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php:
```diff
@@ -0,0 +1,182 @@
+        $hasContent = false;
+        $comment = ' /* ';
+
+        if ($node->hasDefaultValue() || $node->getInfo() !== null || $node->isDeprecated()) {
```
```suggestion
        if ($node->hasDefaultValue() || $node->getInfo() || $node->isDeprecated()) {
```

nicolas-grekas commented on src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php:
```diff
@@ -0,0 +1,182 @@
+
+            if ($node->hasDefaultValue() && !\is_array($defaultValue = $node->getDefaultValue())) {
+                $hasContent = true;
+                $comment .= 'Default value: '.json_encode($defaultValue).'. ';
```
```suggestion
                $comment .= 'Default: '.json_encode($defaultValue).'. ';
```

nicolas-grekas commented on src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php:
```diff
@@ -0,0 +1,182 @@
+    /**
+     * @param self::FORMAT_* $format
+     */
+    public static function generate(ArrayNode $node, string $format = self::FORMAT_PHPDOC): string
```
I wouldn't add a default value, it makes things less explicit

nicolas-grekas commented on src/Symfony/Component/Config/Tests/Builder/Fixtures/ScalarNormalizedTypes/Symfony/Config/ScalarNormalizedTypesConfig.php:
```diff
@@ -128,6 +130,46 @@ public function nested(array $value = []): \Symfony\Config\ScalarNormalizedTypes
+     *     },
+     * } $config
+     */
+    public function configure(#[ArrayShape([
```
let's ensure each argument start on its own line with a starting comment to help the reader:
```suggestion
    public function configure(
        // config:
        #[ArrayShape([
```

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Loader/PhpFileLoader.php:
```diff
@@ -36,13 +37,15 @@
 class PhpFileLoader extends FileLoader
 {
     protected bool $autoRegisterAliasesForSinglyImplementedInterfaces = false;
+    private static array $configurations = [];
```
why static? looks like a bad idea to me :)

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Loader/PhpFileLoader.php:
```diff
@@ -36,13 +37,15 @@
         ?string $env = null,
         private ?ConfigBuilderGeneratorInterface $generator = null,
         bool $prepend = false,
+        private ?ConfigFunctionGenerator $configFunctionGenerator = null,
```
could we find a way that doesn't require changing the constructor? can't we find a way to use the ConfigBuilderGeneratorInterface instance?

nicolas-grekas commented on src/Symfony/Component/Config/Builder/ConfigFunctionGenerator.php:
```diff
@@ -0,0 +1,81 @@
+
+    /**
+     * @param array<string, ConfigurationInterface> $configurations Configurations indexed by their alias.
+     * @return \Closure
```
```suggestion
```

nicolas-grekas commented on src/Symfony/Component/Config/Builder/ConfigFunctionGenerator.php:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    /**
+     * @param array<string, ConfigurationInterface> $configurations Configurations indexed by their alias.
```
```suggestion
     * @param array<string, ConfigurationInterface> $configurations Configurations indexed by their alias
```

nicolas-grekas commented on src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php:
```diff
@@ -0,0 +1,182 @@
+            foreach ($children as $child) {
+                $shape .= \sprintf("%s'%s' => ", str_repeat(' ', $nestingLevel*4), $child->getName());
+                if ($child instanceof ArrayNode) {
+                    $shape .= static::doGenerateJetBrainsArrayShape($child, $nestingLevel+1);
```
```suggestion
                    $shape .= static::doGenerateJetBrainsArrayShape($child, 1 + $nestingLevel);
```

nicolas-grekas commented on src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php:
```diff
@@ -0,0 +1,182 @@
+
+            /** @var NodeInterface $child */
+            foreach ($children as $child) {
+                $shape .= \sprintf("%s'%s' => ", str_repeat(' ', $nestingLevel*4), $child->getName());
```
```suggestion
                $shape .= \sprintf("%s'%s' => ", str_repeat(' ', 4 * $nestingLevel), $child->getName());
```

nicolas-grekas commented on src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php:
```diff
@@ -0,0 +1,182 @@
+                $shape .= ','.static::generateInlinePhpDocForNode($child).PHP_EOL;
+            }
+
+            $shape .= str_repeat(' ', ($nestingLevel-1)*4).']';
```
```suggestion
            $shape .= str_repeat(' ', 4 * ($nestingLevel - 1)).']';
```

nicolas-grekas reviewed (commented):
First quick review

alexandre-daubois commented:
@nicolas-grekas thanks for the review, I addressed all comments. I found a way to keep PhpFileLoader constructor as-is

nicolas-grekas committed:
[HttpFoundation] Remove invalid HTTP method from exception message

staabm committed:
Don't use `die()` in PHPT `--SKIPIF--`

Unlocks a performance optimization in PHPUnit 11.x

see https://staabm.github.io/2024/10/19/phpunit-codesprint-munich.html

fabpot committed:
Update CHANGELOG for 5.4.45

fabpot committed:
Update CONTRIBUTORS for 5.4.45

fabpot committed:
Update VERSION for 5.4.45

fabpot committed:
Merge pull request #58681 from fabpot/release-5.4.45

released v5.4.45

fabpot committed:
Bump Symfony version to 5.4.46

fabpot committed:
Update CHANGELOG for 6.4.13

fabpot committed:
Update VERSION for 6.4.13

fabpot committed:
Merge pull request #58682 from fabpot/release-6.4.13

released v6.4.13

fabpot committed:
Bump Symfony version to 6.4.14

fabpot committed:
Update CHANGELOG for 7.1.6

fabpot committed:
Update VERSION for 7.1.6

fabpot committed:
Merge pull request #58683 from fabpot/release-7.1.6

released v7.1.6

fabpot committed:
Bump Symfony version to 7.1.7

alexandre-daubois commented:
Tests added to cover the `config()` function generation

Valmonzo committed:
[Runtime] Remove unused `SKIPIF` from `dotenv_overload.phpt`

pan93412 committed:
[TwigBridge] Use INTERNAL_VAR_NAME instead of getVarName

nicolas-grekas committed:
bug #58703 [TwigBridge] Use INTERNAL_VAR_NAME instead of getVarName (pan93412)

This PR was squashed before being merged into the 7.2 branch.

Discussion
----------

[TwigBridge] Use INTERNAL_VAR_NAME instead of getVarName

| Q             | A
| ------------- | ---
| Branch?       | 7.2
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | Fix #58702
| License       | MIT

Reapply https://github.com/symfony/symfony/commit/ab116bb5021c03889c3aa0c1c1d6067c7a67740c. This PR is tested and should be worked.

![CleanShot 2024-10-29 at 10 55 17@2x](https://github.com/user-attachments/assets/0639d8b1-a471-4ea0-ae52-b9fd815fd4f8)

Commits
-------

eea0dd3e46e [TwigBridge] Use INTERNAL_VAR_NAME instead of getVarName

ZaneCEO committed:
fix for HttpClientDataCollector fails if proc_open is disabled via php.ini . Closes #58700

xabbuh committed:
minor #58680 [PhpUnitBridge] Don't use `die()` in PHPT `--SKIPIF--` (staabm)

This PR was merged into the 7.2 branch.

Discussion
----------

[PhpUnitBridge] Don't use `die()` in PHPT `--SKIPIF--`

| Q             | A
| ------------- | ---
| Branch?       | 7.2
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Issues        | --
| License       | MIT

Unlocks a performance optimization in PHPUnit 11.5.x

[PHPUnit 11.5.x will be able to avoid subprocess creation](https://github.com/sebastianbergmann/phpunit/pull/5998), when `--SKIPIF--` code in PHPT tests does not `exit()` or `die()` and the logic is side-effect free (output is allowed).

more details in https://staabm.github.io/2024/10/19/phpunit-codesprint-munich.html

Commits
-------

83b09ba453 Don't use `die()` in PHPT `--SKIPIF--`

nicolas-grekas committed:
[Process] On Windows, don't rely on the OS to find executables

nicolas-grekas committed:
[Process] Fix handling empty path found in the PATH env var with ExecutableFinder

alexandre-daubois committed:
[Config] Handle Phar absolute path in `FileLocator`

nicolas-grekas committed:
[HttpClient] Fix Process-based escaping in HttpClientDataCollector

nicolas-grekas committed:
bug #58704 [HttpClient] fix for HttpClientDataCollector fails if proc_open is disabled via php.ini (ZaneCEO)

This PR was merged into the 6.4 branch.

Discussion
----------

[HttpClient] fix for HttpClientDataCollector fails if proc_open is disabled via php.ini

| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | Fix #58700
| License       | MIT

[HttpClientDataCollector::escapePayload](https://github.com/symfony/symfony/blob/7.1/src/Symfony/Component/HttpClient/DataCollector/HttpClientDataCollector.php#L256) creates a `new Process()` -> it fails if `proc_open` is disabled:

> The Process class relies on proc_open, which is not available on your PHP installation.

https://github.com/symfony/symfony/issues/58700

Commits
-------

e28af345004 [HttpClient] Fix Process-based escaping in HttpClientDataCollector
48980a2f05a fix for HttpClientDataCollector fails if proc_open is disabled via php.ini . Closes #58700

nicolas-grekas committed:
[Process] Properly deal with not-found executables on Windows

nicolas-grekas committed:
Merge branch '5.4' into 6.4

* 5.4:
  Bump Symfony version to 5.4.46
  Update VERSION for 5.4.45
  Update CONTRIBUTORS for 5.4.45
  Update CHANGELOG for 5.4.45

nicolas-grekas committed:
Merge branch '6.4' into 7.1

* 6.4:
  [HttpClient] Fix Process-based escaping in HttpClientDataCollector
  fix for HttpClientDataCollector fails if proc_open is disabled via php.ini . Closes #58700
  Bump Symfony version to 6.4.14
  Update VERSION for 6.4.13
  Update CHANGELOG for 6.4.13
  Bump Symfony version to 5.4.46
  Update VERSION for 5.4.45
  Update CONTRIBUTORS for 5.4.45
  Update CHANGELOG for 5.4.45

nicolas-grekas committed:
Merge branch '7.1' into 7.2

* 7.1:
  [HttpClient] Fix Process-based escaping in HttpClientDataCollector
  fix for HttpClientDataCollector fails if proc_open is disabled via php.ini . Closes #58700
  Bump Symfony version to 7.1.7
  Update VERSION for 7.1.6
  Update CHANGELOG for 7.1.6
  Bump Symfony version to 6.4.14
  Update VERSION for 6.4.13
  Update CHANGELOG for 6.4.13
  Bump Symfony version to 5.4.46
  Update VERSION for 5.4.45
  Update CONTRIBUTORS for 5.4.45
  Update CHANGELOG for 5.4.45

nicolas-grekas committed:
bug #58711 [Process] Fix handling empty path found in the PATH env var with ExecutableFinder (nicolas-grekas)

This PR was merged into the 5.4 branch.

Discussion
----------

[Process] Fix handling empty path found in the PATH env var with ExecutableFinder

| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | -
| License       | MIT

Commits
-------

c40ec9cdab6 [Process] Fix handling empty path found in the PATH env var with ExecutableFinder

nicolas-grekas committed:
bug #58723 [Process] Properly deal with not-found executables on Windows (nicolas-grekas)

This PR was merged into the 5.4 branch.

Discussion
----------

[Process] Properly deal with not-found executables on Windows

| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | -
| License       | MIT

Removing the unexpected notices that we see on appveyor:

Before:
![image](https://github.com/user-attachments/assets/d9c788ad-7b29-4b0c-9170-aa08953a5ca2)

After:
![image](https://github.com/user-attachments/assets/a3dc1cc7-ce4c-4fd0-b422-1a6c745bc3d9)

Commits
-------

73140eb44e1 [Process] Properly deal with not-found executables on Windows

nicolas-grekas committed:
Merge branch '5.4' into 6.4

* 5.4:
  [Process] Properly deal with not-found executables on Windows
  [Process] Fix handling empty path found in the PATH env var with ExecutableFinder

nicolas-grekas committed:
Merge branch '6.4' into 7.1

* 6.4:
  [Process] Properly deal with not-found executables on Windows
  [Process] Fix handling empty path found in the PATH env var with ExecutableFinder

nicolas-grekas committed:
Merge branch '7.1' into 7.2

* 7.1:
  [Process] Properly deal with not-found executables on Windows
  [Process] Fix handling empty path found in the PATH env var with ExecutableFinder

nicolas-grekas committed:
minor #58725 [HttpFoundation] Remove invalid HTTP method from exception message (nicolas-grekas)

This PR was merged into the 5.4 branch.

Discussion
----------

[HttpFoundation] Remove invalid HTTP method from exception message

| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Issues        | -
| License       | MIT

This info is not needed so better remove it as its garbage user input anyway.

Commits
-------

596f8bb527e [HttpFoundation] Remove invalid HTTP method from exception message

welcoMattic committed:
Re-add missing Profiler shortcuts on Profiler homepage

knezmilos13 committed:
[Validator] Fix 58691 (missing plural-options in serbian language translation)

Seldaek committed:
[Process] Return built-in cmd.exe commands directly in ExecutableFinder

vltrof committed:
profiler form data collector extart value property if it is setted

derrabus committed:
Merge branch '5.4' into 6.4

* 5.4:
  [HttpFoundation] Remove invalid HTTP method from exception message

derrabus committed:
Merge branch '6.4' into 7.1

* 6.4:
  [HttpFoundation] Remove invalid HTTP method from exception message

derrabus committed:
Merge branch '7.1' into 7.2

* 7.1:
  [HttpFoundation] Remove invalid HTTP method from exception message

alexandre-daubois committed:
[Notifier] Fix test with hard coded date in `SmsboxTransportTest`

alexandre-daubois committed:
[Process] Improve test cleanup by unlinking in a `finally` block

nicolas-grekas committed:
bug #58735 [Process] Return built-in cmd.exe commands directly in ExecutableFinder (Seldaek)

This PR was merged into the 5.4 branch.

Discussion
----------

[Process] Return built-in cmd.exe commands directly in ExecutableFinder

| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | Refs #58734
| License       | MIT

Commits
-------

3b144a3a243 [Process] Return built-in cmd.exe commands directly in ExecutableFinder

xabbuh committed:
fix the directory separator being used

xabbuh committed:
ignore case of built-in cmd.exe commands

nicolas-grekas committed:
minor #58748 [Process] ignore case of built-in cmd.exe commands (xabbuh)

This PR was merged into the 5.4 branch.

Discussion
----------

[Process] ignore case of built-in cmd.exe commands

| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Issues        |
| License       | MIT

Commits
-------

278cabdcef5 ignore case of built-in cmd.exe commands

nicolas-grekas committed:
minor #58746 [Process] Improve test cleanup by unlinking in a `finally` block (alexandre-daubois)

This PR was merged into the 5.4 branch.

Discussion
----------

[Process] Improve test cleanup by unlinking in a `finally` block

| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Issues        | -
| License       | MIT

Follwup for #58711, I realized afterwards that we may unlink in `finally` block to ensure proper test clean up in case something goes wrong.

Commits
-------

5c547c958b9 [Process] Improve test cleanup by unlinking in a `finally` block

nicolas-grekas committed:
minor #58747 [Process] fix the directory separator being used (xabbuh)

This PR was merged into the 5.4 branch.

Discussion
----------

[Process] fix the directory separator being used

| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Issues        |
| License       | MIT

Commits
-------

2526495eb75 fix the directory separator being used

xabbuh committed:
Merge branch '5.4' into 6.4

* 5.4:
  fix the directory separator being used
  ignore case of built-in cmd.exe commands
  [Process] Improve test cleanup by unlinking in a `finally` block
  [Process] Return built-in cmd.exe commands directly in ExecutableFinder

xabbuh committed:
fix the path separator being used

nicolas-grekas committed:
minor #58750 [Process] fix the path separator being used (xabbuh)

This PR was merged into the 5.4 branch.

Discussion
----------

[Process] fix the path separator being used

| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Issues        |
| License       | MIT

I missed to also update the assertion in #58747.

Commits
-------

c24ebca7c44 fix the path separator being used

xabbuh committed:
fix the constant being used

nicolas-grekas committed:
minor #58751 [Process] fix the constant being used (xabbuh)

This PR was merged into the 5.4 branch.

Discussion
----------

[Process] fix the constant being used

| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Issues        |
| License       | MIT

Commits
-------

b811c1a224d fix the constant being used

xabbuh committed:
Merge branch '5.4' into 6.4

* 5.4:
  fix the constant being used
  fix the path separator being used

nicolas-grekas committed:
[Process] Fix escaping /X arguments on Windows

nicolas-grekas committed:
bug #58752 [Process] Fix escaping /X arguments on Windows (nicolas-grekas)

This PR was merged into the 5.4 branch.

Discussion
----------

[Process] Fix escaping /X arguments on Windows

| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | Fix #58733
| License       | MIT

Commits
-------

16902ec540c [Process] Fix escaping /X arguments on Windows

nicolas-grekas committed:
minor #58745 [Notifier] Fix test with hard coded date in `SmsboxTransportTest` (alexandre-daubois)

This PR was merged into the 7.1 branch.

Discussion
----------

[Notifier] Fix test with hard coded date in `SmsboxTransportTest`

| Q             | A
| ------------- | ---
| Branch?       | 7.1
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | -
| License       | MIT

Fixes CI, see https://github.com/symfony/symfony/actions/runs/11654514385/job/32448126831#step:8:1756.

Commits
-------

2ee5364e3a6 [Notifier] Fix test with hard coded date in `SmsboxTransportTest`

nicolas-grekas committed:
bug #58739 [WebProfilerBoundle] form data collector check passed and resolved options are defined (vltrof)

This PR was submitted for the 6.4 branch but it was merged into the 5.4 branch instead.

Discussion
----------

[WebProfilerBoundle] form data collector check passed and resolved options are defined

| Q             | A
| ------------- | ---
| Branch?      | 6.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | Fix #58665
| License       | MIT

Greetings!
When passed or resolved value casted to false (null/false) it sets wrong value what cause exceptions described in issue. Explicit check if value defined prevent this behavior
Maybe codestyle should be fixed because of ternary

Commits
-------

93faa963eb6 profiler form data collector extart value property if it is setted

nicolas-grekas committed:
minor #58731 [Validator] Fix 58691 (missing plural-options in serbian language translation) (knezmilos13)

This PR was squashed before being merged into the 5.4 branch.

Discussion
----------

[Validator] Fix 58691 (missing plural-options in serbian language translation)

| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | Fix #58691
| License       | MIT

Pull request contains two commits that do the following:

1. Fix Serbian translations - wrong number of plural options that was causing a crash, see https://github.com/symfony/symfony/issues/58691
2. General improvements of Serbian translation:
    * Serbian-latin and Serbian-cyrillic translations should use the exact same wording, only the alphabet used should differ
    * Some minor changes to make the translation more consistent

Commits
-------

1e071cbc3c2 [Validator] Fix 58691 (missing plural-options in serbian language translation)

nicolas-grekas committed:
bug #58728 [WebProfilerBundle] Re-add missing Profiler shortcuts on Profiler homepage (welcoMattic)

This PR was merged into the 6.4 branch.

Discussion
----------

[WebProfilerBundle] Re-add missing Profiler shortcuts on Profiler homepage

| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | yes
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Issues        | Fix #58693 <!-- prefix each issue number with "Fix #", no need to create an issue if none exists, explain below instead -->
| License       | MIT

Before Profiler redesign, on "homepage" (`/_profiler` path), there was sidebar shortcuts ("Last 10", "Latest", and "Search").
![Screenshot 2024-10-31 at 15-04-50 Symfony Profiler](https://github.com/user-attachments/assets/018847dd-c67c-41ae-a62d-4975904bf478)

Since this redesign, as mentioned in the related issue, shortcuts have disappeared
![Screenshot 2024-10-31 at 15-05-22 Symfony Profiler](https://github.com/user-attachments/assets/eb58515d-839d-4675-8c1a-761f55a8bcb8)

I've fixed it on this PR, it works exactly like before the redesign

![Screenshot 2024-10-31 at 15-05-46 Symfony Profiler](https://github.com/user-attachments/assets/fddcdf3e-01e8-4b39-921b-0d998e5f1231)

Commits
-------

4c6e6ebc7e4 Re-add missing Profiler shortcuts on Profiler homepage

nicolas-grekas committed:
bug #58713 [Config] Handle Phar absolute path in `FileLocator` (alexandre-daubois)

This PR was merged into the 5.4 branch.

Discussion
----------

[Config] Handle Phar absolute path in `FileLocator`

| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | Fix #58708
| License       | MIT

Commits
-------

4f00d6f8531 [Config] Handle Phar absolute path in `FileLocator`

nicolas-grekas committed:
minor #58695 [Runtime] Remove unused `SKIPIF` from `dotenv_overload.phpt` (Valmonzo)

This PR was merged into the 6.4 branch.

Discussion
----------

[Runtime] Remove unused `SKIPIF` from `dotenv_overload.phpt`

| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Issues        |
| License       | MIT

Since`Runtime` requires `Dotenv` 5.4 version, this condition is useless because `Dotenv` introduced a fourth parameter in 5.4

Commits
-------

3c69ee7810a [Runtime] Remove unused `SKIPIF` from `dotenv_overload.phpt`

xabbuh committed:
Merge branch '6.4' into 7.1

* 6.4:
  fix the constant being used
  fix the path separator being used
  fix the directory separator being used
  ignore case of built-in cmd.exe commands
  [Process] Improve test cleanup by unlinking in a `finally` block
  [Process] Return built-in cmd.exe commands directly in ExecutableFinder
  Re-add missing Profiler shortcuts on Profiler homepage

nicolas-grekas committed:
Merge branch '5.4' into 6.4

* 5.4:
  [Validator] Fix 58691 (missing plural-options in serbian language translation)
  profiler form data collector extart value property if it is setted
  [Process] Fix escaping /X arguments on Windows
  [Config] Handle Phar absolute path in `FileLocator`

nicolas-grekas committed:
Merge branch '6.4' into 7.1

* 6.4:
  [Validator] Fix 58691 (missing plural-options in serbian language translation)
  profiler form data collector extart value property if it is setted
  [Process] Fix escaping /X arguments on Windows
  [Config] Handle Phar absolute path in `FileLocator`
  [Runtime] Remove unused `SKIPIF` from `dotenv_overload.phpt`

nicolas-grekas committed:
Merge branch '7.1' into 7.2

* 7.1:
  [Validator] Fix 58691 (missing plural-options in serbian language translation)
  profiler form data collector extart value property if it is setted
  [Process] Fix escaping /X arguments on Windows
  fix the constant being used
  fix the path separator being used
  fix the directory separator being used
  ignore case of built-in cmd.exe commands
  [Process] Improve test cleanup by unlinking in a `finally` block
  [Notifier] Fix test with hard coded date in `SmsboxTransportTest`
  [Process] Return built-in cmd.exe commands directly in ExecutableFinder
  Re-add missing Profiler shortcuts on Profiler homepage
  [Config] Handle Phar absolute path in `FileLocator`
  [Runtime] Remove unused `SKIPIF` from `dotenv_overload.phpt`

nicolas-grekas committed:
minor #58710 [Process] On Windows, don't rely on the OS to find executables (nicolas-grekas)

This PR was merged into the 7.2 branch.

Discussion
----------

[Process] On Windows, don't rely on the OS to find executables

| Q             | A
| ------------- | ---
| Branch?       | 7.2
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Issues        | -
| License       | MIT

Porting part of https://github.com/composer/composer/pull/12180 here:

On Windows, when searching for an executable, the OS always looks at the current directory before using the PATH variable. This makes it easier than desired to hijack executables. Unix-like OSes don't have this issue.

This PR proposes to rely on ExecutableFinder instead.

Commits
-------

b35a7d42931 [Process] On Windows, don't rely on the OS to find executables

nicolas-grekas committed:
minor #58585 Rename annotations to attributes in `AttributeLoader` (alamirault)

This PR was merged into the 7.2 branch.

Discussion
----------

Rename annotations to attributes in `AttributeLoader`

| Q             | A
| ------------- | ---
| Branch?       | 7.2
| Bug fix?      | no
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Issues        | Fix #... <!-- prefix each issue number with "Fix #", no need to create an issue if none exists, explain below instead -->
| License       | MIT

Annotations are not supported anymore in AttributeLoader, so we can rename some variables

Commits
-------

9ab0752155b [Serializer] Rename annotations to attributes in AttributeLoader

nicolas-grekas committed:
minor #58584 [Routing] Rename annotations to attribute in `AttributeClassLoader` (alamirault)

This PR was squashed before being merged into the 7.2 branch.

Discussion
----------

[Routing] Rename annotations to attribute in `AttributeClassLoader`

| Q             | A
| ------------- | ---
| Branch?       | 7.2
| Bug fix?      | no
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Issues        | Fix #... <!-- prefix each issue number with "Fix #", no need to create an issue if none exists, explain below instead -->
| License       | MIT

Annotations are not supported anymore in `AttributeClassLoader`, so we can rename some methods

Commits
-------

8fd78bd29f9 [Routing] Rename annotations to attribute in `AttributeClassLoader`

javiereguiluz committed:
[Mime] Don't require passig the encoder name to TextPart

nicolas-grekas committed:
bug #58764 [Mime] Don't require passing the encoder name to `TextPart` (javiereguiluz)

This PR was merged into the 7.2 branch.

Discussion
----------

[Mime] Don't require passing the encoder name to `TextPart`

| Q             | A
| ------------- | ---
| Branch?       | 7.2
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Issues        | -
| License       | MIT

While preparing this blog post: https://symfony.com/blog/new-in-symfony-7-2-mime-improvements

I realized that it's unnecessary to pass the encoder name as the first argument of `TextPart::addEncoder()` because the `ContentEncoderInterface` already requires defining a `getName()` method.

Commits
-------

2ce04c8824b [Mime] Don't require passig the encoder name to TextPart

alexandre-daubois commented:
https://github.com/symfony/symfony/pull/58771
